cmd/deptool/analyze.go

```diff
@@ -62,87 +69,277 @@ func analyzedDependency(depInfo projectDependency) analyzedProjectDependency {
+		return nil
+	})
+	if err != nil && err != storer.ErrStop {
+		fmt.Printf("unable to iterate on log entries : %v\n", err)
 		exitErr()
 	}
-	parentBranchName := parentBranch.String()
+	if revCommit == nil {
+		fmt.Printf("the commit object for short commit %s was missing ?!\n", lookoutCommit)
+		exitErr()
+	}
+	cIter.Close()
+	return revCommit, nil
+}
+
+func getBranches(repo *git.Repository) (branches map[plumbing.Hash]*plumbing.Reference, err error) {
+	remoteOrigin, err := repo.Remote("origin")
+	if err != nil {
+		fmt.Printf("unable to retrieve origin remote : %v\n", err)
+		return nil, err
+	}
 
-	return analyzedProjectDependency{
-		projectDependency:      depInfo,
-		branchName:             parentBranchName,
-		fullCommitHash:         revCommit.Hash.String(),
-		latestBranchCommit:     parentBranch.Hash().String(),
-		latestBranchCommitTime: branchTime[parentBranch].UTC(),
+	remoteRefs, err := remoteOrigin.List(&git.ListOptions{})
+	if err != nil {
+		fmt.Printf("unable to list remote refs : %v\n", err)
+		return nil, err
+	}
+	branchPrefix := "refs/heads/"
+	branches = make(map[plumbing.Hash]*plumbing.Reference, 0)
+	for _, remoteRef := range remoteRefs {
+		refName := remoteRef.Name().String()
+		if !strings.HasPrefix(refName, branchPrefix) {
+			continue
+		}
+		branches[remoteRef.Hash()] = remoteRef
+	}
+	return branches, nil
+}
+
+func findLatestVersion(repo *git.Repository, latestCommitRef *plumbing.Reference, revCommit *object.Commit, pkgName string) (updatedLatestCommit *object.Commit, version string, workspaceVersion bool, err error) {
+	// create a list of all the commits between the head and the current.
+	commits := []*object.Commit{}
+	headCommit, err := repo.CommitObject(latestCommitRef.Hash())
+	if err != nil {
+		return nil, "", false, err
+	}
+	for {
+		commits = append(commits, headCommit)
+		if headCommit.Hash == revCommit.Hash {
+			// we're done.
+			break
+		}
+		if parent, err := headCommit.Parent(0); err != nil || parent == nil {
+			break
+		} else {
+			headCommit = parent
+		}
+	}
+
+	var versions []string
+	var workspaceVer []bool
+	for _, commit := range commits {
+		version, workspaceVersion, err := findCargoVersionForCommit(repo, pkgName, commit)
+		if err != nil {
+			return nil, "", false, err
+		}
+		versions = append(versions, version)
+		workspaceVer = append(workspaceVer, workspaceVersion)
+	}
+	for i := 1; i < len(versions); i++ {
+		if versions[i] != versions[i-1] {
+			// the version at i-1 is "newer", so we should pick that one.
+			return commits[i-1], versions[i-1], workspaceVer[i-1], nil
+		}
+	}
+
+	return commits[len(commits)-1], versions[len(commits)-1], workspaceVer[len(commits)-1], nil
+}
+
+func findCargoVersionForCommit(repo *git.Repository, pkgName string, commit *object.Commit) (string, bool, error) {
+	treeRoot, err := commit.Tree()
+	if err != nil {
+		return "", false, err
+	}
+	rootCargoFile, err := treeRoot.File("Cargo.toml")
+	internalWorkspacePackage := false
+	if err == nil {
```

Comment:
I would check for the error instead, to avoid the nesting. It will make the function much more readable.

Reply:
You could do

```
if err != nil && !internalWorkspacePackage {
 // print/return error
}

// handle it without error
```

Alternatively you can move the existing if blocks into functions of their own, since they are pretty large.